src/iconid.rs

```diff
@@ -0,0 +1,157 @@
+//! Identification of icons and resolution of glyph ids. Assumes Google style icon font input.
+use skrifa::{
+    charmap::Charmap,
+    raw::{
+        tables::gsub::{ExtensionSubtable, LigatureSubstFormat1, SubstitutionLookup},
+        FontRef, TableProvider,
+    },
+    GlyphId,
+};
+use smol_str::SmolStr;
+
+use crate::error::IconResolutionError;
+
+#[derive(Clone, Debug)]
+pub enum IconIdentifier {
+    GlyphId(GlyphId),
+    Codepoint(u32),
+    Name(SmolStr),
+}
+
+impl IconIdentifier {
+    pub fn resolve(&self, font: &FontRef) -> Result<GlyphId, IconResolutionError> {
+        match self {
+            IconIdentifier::GlyphId(gid) => Ok(*gid),
+            IconIdentifier::Codepoint(cp) => font
+                .cmap()
+                .map_err(IconResolutionError::ReadError)?
+                .map_codepoint(*cp)
+                .ok_or(IconResolutionError::NoCmapEntry(*cp)),
+            IconIdentifier::Name(name) => resolve_icon_ligature(font, name.as_str()),
+        }
+    }
+}
+
+fn resolve_ligature(
+    liga: &LigatureSubstFormat1<'_>,
+    text: &str,
+    gids: &[GlyphId],
+) -> Result<Option<GlyphId>, IconResolutionError> {
+    let Some(first) = gids.first() else {
+        return Err(IconResolutionError::NoGlyphIds(text.to_string()));
+    };
+    let coverage = liga.coverage().map_err(IconResolutionError::ReadError)?;
+    let Some(set_index) = coverage.get(*first) else {
+        return Ok(None);
+    };
+    let set = liga
+        .ligature_sets()
+        .get(set_index as usize)
+        .map_err(IconResolutionError::ReadError)?;
+    // Seek a ligature that matches glyphs 2..N of name
+    // We don't care about speed
+    let gids = &gids[1..];
+    for liga in set.ligatures().iter() {
+        let liga = liga.map_err(IconResolutionError::ReadError)?;
+        if liga.component_count() as usize != gids.len() + 1 {
+            continue;
+        }
+        if gids
+            .iter()
+            .zip(liga.component_glyph_ids())
+            .all(|(gid, component)| *gid == component.get())
+        {
+            return Ok(Some(liga.ligature_glyph())); // We found it!
+        }
+    }
+    Ok(None)
+}
+
+pub fn resolve_icon_ligature(font: &FontRef, name: &str) -> Result<GlyphId, IconResolutionError> {
+    let charmap = Charmap::new(font);
+    let gids = name
+        .chars()
+        .map(|c| {
+            charmap
+                .map(c)
+                .ok_or(IconResolutionError::UnmappedCharError(c))
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // Step 1: try to find a ligature that starts with our first gid
+    let gsub = font.gsub().map_err(IconResolutionError::ReadError)?;
+    let lookups = gsub.lookup_list().map_err(IconResolutionError::ReadError)?;
+    for lookup in lookups.lookups().iter() {
+        let lookup = lookup.map_err(IconResolutionError::ReadError)?;
+        match lookup {
```

Comment:
instead of matching the lookup directly, try calling the 'subtables' method and matching on that; this will handle extensions for you:

https://docs.rs/read-fonts/latest/read_fonts/tables/gsub/enum.SubstitutionLookup.html#method.subtables

Reply:
That'll be much nicer. Good catch.

Reply:
Dramatically better!

Reply:
yea we added that convenience a couple months ago after bumping into similar headaches in a various other places.